File: packages/server/pkg/flowgraph/layout.go
package flowgraph

import (
	"fmt"
	"sort"

	"github.com/the-dev-tools/dev-tools/packages/server/pkg/idwrap"
	"github.com/the-dev-tools/dev-tools/packages/server/pkg/model/mflow"
)

// Layout computes node positions using BFS-based level assignment.
// Each node's level is max(parent_levels) + 1, ensuring proper dependency ordering.
func Layout(nodes []mflow.Node, edges []mflow.Edge, startNodeID idwrap.IDWrap, config LayoutConfig) (*LayoutResult, error) {
	if len(nodes) == 0 {
		return &LayoutResult{
			Positions: make(map[idwrap.IDWrap]Position),
			Levels:    make(map[idwrap.IDWrap]int),
			MaxLevel:  0,
		}, nil
	}

	// Build adjacency lists
	outgoingEdges := BuildOutgoingAdjacency(edges)
	incomingEdges := BuildIncomingAdjacency(edges)

	// Calculate dependency levels using BFS
	nodeLevels := make(map[idwrap.IDWrap]int)
	levelNodes := make(map[int][]idwrap.IDWrap)

	// Start BFS from start node
	queue := []idwrap.IDWrap{startNodeID}
	nodeLevels[startNodeID] = 0
	levelNodes[0] = []idwrap.IDWrap{startNodeID}

	// Safety counter to prevent infinite loops on cyclic graphs
	processedCount := 0
	maxProcessed := len(nodes) * len(nodes)
	if maxProcessed < 10000 {
		maxProcessed = 10000
	}

	for len(queue) > 0 {
		if processedCount > maxProcessed {
			break
		}
		processedCount++

		currentNodeID := queue[0]
		queue = queue[1:]

		// Process all children
		for _, childID := range outgoingEdges[currentNodeID] {
			// Calculate the maximum level of all parents + 1
			maxParentLevel := -1
			for _, parentID := range incomingEdges[childID] {
				if parentLevel, exists := nodeLevels[parentID]; exists {
					if parentLevel > maxParentLevel {
						maxParentLevel = parentLevel
					}
				}
			}

			childLevel := maxParentLevel + 1

			// Only update if this is a new node or we found a deeper level
			if existingLevel, exists := nodeLevels[childID]; !exists || childLevel > existingLevel {
				// Remove from old level if it existed
				if exists {
					oldLevelNodes := levelNodes[existingLevel]
					for i, nodeID := range oldLevelNodes {
						if nodeID == childID {
							levelNodes[existingLevel] = append(oldLevelNodes[:i], oldLevelNodes[i+1:]...)
							break
						}
					}
				}

				// Add to new level
				nodeLevels[childID] = childLevel
				levelNodes[childLevel] = append(levelNodes[childLevel], childID)
				queue = append(queue, childID)
			}
		}
	}

	// Find max level
	maxLevel := 0
	for level := range levelNodes {
		if level > maxLevel {
			maxLevel = level
		}
	}

	// Calculate positions based on orientation
	positions := make(map[idwrap.IDWrap]Position)

	for level := 0; level <= maxLevel; level++ {
		nodesAtLevel := levelNodes[level]
		if len(nodesAtLevel) == 0 {
			continue
		}

		// Calculate primary axis position (depth direction)
		primaryPos := config.StartX
		if config.Orientation == LayoutVertical {
			primaryPos = config.StartY
		}
		primaryPos += float64(level) * config.SpacingPrimary

		// Calculate secondary axis positions (centered around start)
		totalSecondary := float64((len(nodesAtLevel) - 1)) * config.SpacingSecondary
		startSecondary := config.StartY
		if config.Orientation == LayoutVertical {
			startSecondary = config.StartX
		}
		startSecondary -= totalSecondary / 2

		for i, nodeID := range nodesAtLevel {
			secondaryPos := startSecondary + float64(i)*config.SpacingSecondary

			var pos Position
			if config.Orientation == LayoutHorizontal {
				pos = Position{X: primaryPos, Y: secondaryPos}
			} else {
				pos = Position{X: secondaryPos, Y: primaryPos}
			}
			positions[nodeID] = pos
		}
	}

	return &LayoutResult{
		Positions: positions,
		Levels:    nodeLevels,
		MaxLevel:  maxLevel,
	}, nil
}

// ApplyLayout applies the layout result to a slice of nodes.
func ApplyLayout(nodes []mflow.Node, result *LayoutResult) {
	for i := range nodes {
		if pos, ok := result.Positions[nodes[i].ID]; ok {
			nodes[i].PositionX = pos.X
			nodes[i].PositionY = pos.Y
		}
	}
}

// ApplyLayoutToNodePtrs applies the layout result to a map of node pointers.
func ApplyLayoutToNodePtrs(nodeMap map[idwrap.IDWrap]*mflow.Node, result *LayoutResult) {
	for nodeID, pos := range result.Positions {
		if node, ok := nodeMap[nodeID]; ok {
			node.PositionX = pos.X
			node.PositionY = pos.Y
		}
	}
}

// LinearizeNodes returns nodes in BFS traversal order (for YAML export).
// Neighbors are sorted alphabetically by name for deterministic ordering.
// Disconnected nodes are appended at the end, also sorted alphabetically.
func LinearizeNodes(startNodeID idwrap.IDWrap, allNodes []mflow.Node, edges []mflow.Edge) []mflow.Node {
	if len(allNodes) == 0 {
		return nil
	}

	// Build node map for quick lookup
	nodeMap := make(map[idwrap.IDWrap]mflow.Node)
	for _, n := range allNodes {
		nodeMap[n.ID] = n
	}

	// Build edges by source
	edgesBySource := BuildOutgoingAdjacency(edges)

	// BFS traversal
	visited := make(map[idwrap.IDWrap]bool)
	var result []mflow.Node
	queue := []idwrap.IDWrap{startNodeID}
	visited[startNodeID] = true

	for len(queue) > 0 {
		currentID := queue[0]
		queue = queue[1:]

		if n, ok := nodeMap[currentID]; ok {
			result = append(result, n)
		}

		// Get all outgoing edges from current node
		targetIDs := edgesBySource[currentID]
		var neighbors []mflow.Node

		for _, targetID := range targetIDs {
			if target, ok := nodeMap[targetID]; ok {
				neighbors = append(neighbors, target)
			}
		}

		// Sort neighbors alphabetically by name for deterministic ordering
		sort.Slice(neighbors, func(i, j int) bool {
			return neighbors[i].Name < neighbors[j].Name
		})

		// Add unvisited neighbors to queue
		for _, neighbor := range neighbors {
			if !visited[neighbor.ID] {
				visited[neighbor.ID] = true
				queue = append(queue, neighbor.ID)
			}
		}
	}

	// Handle disconnected nodes (not reachable from start)
	var disconnected []mflow.Node
	for _, n := range allNodes {
		if !visited[n.ID] {
			disconnected = append(disconnected, n)
		}
	}

	// Sort disconnected nodes alphabetically
	sort.Slice(disconnected, func(i, j int) bool {
		return disconnected[i].Name < disconnected[j].Name
	})

	// Append disconnected nodes to result
	result = append(result, disconnected...)

	return result
}

// LayoutNodes is a convenience function that performs layout and applies positions.
// It returns an error if the start node is not found.
func LayoutNodes(nodes []mflow.Node, edges []mflow.Edge, config LayoutConfig) error {
	startNode, found := FindStartNode(nodes)
	if !found {
		return fmt.Errorf("start node not found")
	}

	result, err := Layout(nodes, edges, startNode.ID, config)
	if err != nil {
		return err
	}

	ApplyLayout(nodes, result)
	return nil
}

File: packages/client/src/features/agent/layout.ts
import type { EdgeInfo, NodeInfo } from './types';

export type LayoutOrientation = 'horizontal' | 'vertical';

export interface LayoutConfig {
  orientation: LayoutOrientation;
  spacingPrimary: number;
  spacingSecondary: number;
  startX: number;
  startY: number;
}

export interface Position {
  x: number;
  y: number;
}

export interface LayoutResult {
  levels: Map<string, number>;
  maxLevel: number;
  positions: Map<string, Position>;
}

export const defaultHorizontalConfig = (): LayoutConfig => ({
  orientation: 'horizontal',
  spacingPrimary: 300,
  spacingSecondary: 150,
  startX: 0,
  startY: 0,
});

export const defaultVerticalConfig = (): LayoutConfig => ({
  orientation: 'vertical',
  spacingPrimary: 300,
  spacingSecondary: 400,
  startX: 0,
  startY: 0,
});

const buildOutgoingAdjacency = (edges: EdgeInfo[]): Map<string, string[]> => {
  const adj = new Map<string, string[]>();
  for (const e of edges) {
    const existing = adj.get(e.sourceId) ?? [];
    existing.push(e.targetId);
    adj.set(e.sourceId, existing);
  }
  return adj;
};

const findStartNode = (nodes: NodeInfo[]): NodeInfo | undefined => nodes.find((n) => n.kind === 'ManualStart');

/**
 * Layout computes node positions using BFS-based level assignment.
 * Each node's level is max(parent_levels) + 1, ensuring proper dependency ordering.
 * Cycles are handled by only visiting each node once.
 */
export const layout = (
  nodes: NodeInfo[],
  edges: EdgeInfo[],
  startNodeId: string,
  config: LayoutConfig,
): LayoutResult => {
  if (nodes.length === 0) {
    return {
      levels: new Map(),
      maxLevel: 0,
      positions: new Map(),
    };
  }

  const outgoingEdges = buildOutgoingAdjacency(edges);

  const nodeLevels = new Map<string, number>();
  const levelNodes = new Map<number, string[]>();
  const visited = new Set<string>();

  // Start BFS from start node
  const queue: string[] = [startNodeId];
  nodeLevels.set(startNodeId, 0);
  levelNodes.set(0, [startNodeId]);
  visited.add(startNodeId);

  while (queue.length > 0) {
    const currentNodeId = queue.shift()!;
    const currentLevel = nodeLevels.get(currentNodeId) ?? 0;

    // Process all children
    const children = outgoingEdges.get(currentNodeId) ?? [];
    for (const childId of children) {
      // Skip if already visited (handles cycles)
      if (visited.has(childId)) continue;

      // Child level is parent level + 1
      const childLevel = currentLevel + 1;

      // Mark as visited and assign level
      visited.add(childId);
      nodeLevels.set(childId, childLevel);
      const currentLevelNodes = levelNodes.get(childLevel) ?? [];
      currentLevelNodes.push(childId);
      levelNodes.set(childLevel, currentLevelNodes);
      queue.push(childId);
    }
  }

  // Find max level
  let maxLevel = 0;
  for (const level of levelNodes.keys()) {
    if (level > maxLevel) maxLevel = level;
  }

  // Calculate positions based on orientation
  const positions = new Map<string, Position>();

  for (let level = 0; level <= maxLevel; level++) {
    const nodesAtLevel = levelNodes.get(level) ?? [];
    if (nodesAtLevel.length === 0) continue;

    // Calculate primary axis position (depth direction)
    let primaryPos = config.orientation === 'horizontal' ? config.startX : config.startY;
    primaryPos += level * config.spacingPrimary;

    // Calculate secondary axis positions (centered around start)
    const totalSecondary = (nodesAtLevel.length - 1) * config.spacingSecondary;
    let startSecondary = config.orientation === 'horizontal' ? config.startY : config.startX;
    startSecondary -= totalSecondary / 2;

    for (let i = 0; i < nodesAtLevel.length; i++) {
      const nodeId = nodesAtLevel[i]!;
      const secondaryPos = startSecondary + i * config.spacingSecondary;

      const pos: Position =
        config.orientation === 'horizontal' ? { x: primaryPos, y: secondaryPos } : { x: secondaryPos, y: primaryPos };

      positions.set(nodeId, pos);
    }
  }

  return {
    levels: nodeLevels,
    maxLevel,
    positions,
  };
};

/**
 * layoutNodes is a convenience function that finds the start node and performs layout.
 * Returns null if no start node is found.
 */
export const layoutNodes = (
  nodes: NodeInfo[],
  edges: EdgeInfo[],
  config: LayoutConfig = defaultHorizontalConfig(),
): LayoutResult | null => {
  const startNode = findStartNode(nodes);
  if (!startNode) return null;

  return layout(nodes, edges, startNode.id, config);
};
